### PR TITLE
Guide to Spring Cloud Kubernetes

### LINKED ISSUE #2592: Spring Cloud Config Server on Kubernetes (Part 1)
We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner. Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for an articles to be reviewed and/or published.

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

### Proposed title of article
Spring Cloud Config Server on Kubernetes (Part 1)

### Introduction paragraph (2-3 paragraphs):

This is the first of a two part article where I’ll show you how to use Spring Cloud Config Server to build centralised configuration for your microservices.  In this first post we’ll create a Config Service that pulls its data from a Git repo and exposes it to other services. We’ll also create a consumer, a Config Consumer Service that pulls from the Config Service on startup.  The diagram below shows how the services will interact.

In part two of this article we’ll take the two services and deploy them to Kubernetes. We’ll initially deploy and test on a local Kubernetes cluster, before deploying on Azures manged Kubernetes Service, AKS.

### Key takeaways: 
1.Creating the Config Service
2.Service Configuration
3.Defining a Dockerfile
4.Config Consumer Service

### References: 
Please list links to any published content/research that you intend to use to support/guide this article.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by odiwuoramos**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
This fixes #2592

### CODE REVIEW

**Comment by hectorkambow** [comment]
@odiwuoramos 
Good afternoon and thank you for submitting your article for publishing.
Your article has been entered into our queue and should be reviewed (and published pending any changes) as soon as a peer reviewer is finished with the ones in the queue before it. 🚀

Thank you.

**Comment by ahmadmardeni1** [comment]
@Ericgacoki Will review this PR.

**Comment by ahmadmardeni1** [comment]
I just noticed that we didn't approve your topic-suggestion, you can't add an article unless we approve it. @odiwuoramos

**Comment by odiwuoramos** [comment]
> I just noticed that we didn't approve your topic-suggestion, you can't add an article unless we approve it. @odiwuoramos

Noted!

**Comment by ahmadmardeni1** [comment]
We will accept this topic for now - for the future please make sure not to add an article unless it has the topic-approved label. @odiwuoramos 

Re-opened and approved [your topic-suggestion](https://github.com/section-engineering-education/engineering-education/issues/2592) - I noticed that it has "part 1" in the title. If you are willing to write parts 1 and 2, then I feel your topic does not need more than 1 part. 

Let me know if your article is fully ready to be reviewed or still needs a part 2 (Then you have to re-write this one and make it just 1 part). @odiwuoramos

**Comment by odiwuoramos** [comment]
> We will accept this topic for now - for the future please make sure not to add an article unless it has the topic-approved label. @odiwuoramos
> 
> Re-opened and approved [your topic-suggestion](https://github.com/section-engineering-education/engineering-education/issues/2592) - I noticed that it has "part 1" in the title. If you are willing to write parts 1 and 2, then I feel your topic does not need more than 1 part.
> 
> Let me know if your article is fully ready to be reviewed or still needs a part 2 (Then you have to re-write this one and make it just 1 part). @odiwuoramos

It's ready for review! **No part 2**

**Comment by ahmadmardeni1** [comment]
Please start the review @Ericgacoki

File: content/articles/spring-cloud-k8/index.md [hunk 1]
```diff
@@ -0,0 +1,234 @@
+### Tutorial overview
+
+In this tutorial, I'll show you how to use Kubernetes and Spring cloud to build your microservice application. It'll be an interesting project that will walk you through the whole process of integrating Kubernetes with Spring Boot applications.
```
Comment (Ericgacoki) [inline content/articles/spring-cloud-k8/index.md:3]:
```suggestion
### Tutorial overview
In this tutorial, we will learn how to use Kubernetes and Spring cloud to build a micro-service application and how to integrate it with Spring Boot.
```
Please remove the blank line(s) between a heading and the paragraph that follows it. Do this all through the article.

**Review by Ericgacoki — CHANGES_REQUESTED** [review]
Hello @odiwuoramos, kindly resolve the following issue before I proceed with the review.

File: content/articles/spring-cloud-k8/index.md (hunk 1 above)
Comment (odiwuoramos) [reply content/articles/spring-cloud-k8/index.md]: @Ericgacoki ,fixed.

File: content/articles/spring-cloud-k8/index.md [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+```
+
+Output:  
+
+```bash
+secret/db-secret created
+```
+
+Now proceed and add the following configurations in our deployment `YAML` file.  
+
```
Comment (Ericgacoki) [inline content/articles/spring-cloud-k8/index.md:116]:
```suggestion

Proceed to add the following configurations in our deployment `YAML` file.   

```

File: content/articles/spring-cloud-k8/index.md [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+                key: username
+          - name: MONGO_INITDB_ROOT_PASSWORD
+            valueFrom:
+              secretKeyRef:
+              # this password is retrieved from the previous section
+                name: db-secret
+                key: password
+```
+
+This deployment configuration file creates a `mongo: latest image. It then sets both of its username and password we created previously with a default`admin database.
```
Comment (Ericgacoki) [inline content/articles/spring-cloud-k8/index.md:155]:
```suggestion
This deployment configuration file creates a `mongo: latest image`. It also sets the username and password that we created earlier with a default admin database.
```

File: content/articles/spring-cloud-k8/index.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+# specifying the port to use
+spring.data.mongodb.port=27017
+# The database name here is defaulted to admin
+spring.data.mongodb.database=admin
+# for security, we hide the username and password as they are sensitive
+spring.data.mongodb.username=${MONGO_USERNAME}
+spring.data.mongodb.password=${MONGO_PASSWORD}
+```
+
+Now that we have got a complete setup of how this application works, clone the complete code on <https://replit.com/@odiwuoramos/spring-cloud> and test the application on your local machine or as well play with the code on the hosting platform.
```
Comment (Ericgacoki) [inline content/articles/spring-cloud-k8/index.md:176]:
```suggestion
Now that we have understood how this application works, clone the [complete code](https://replit.com/@odiwuoramos/spring-cloud) and test it locally on your machine or play around with the code on the hosting platform.
```

File: content/articles/spring-cloud-k8/index.md [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+### Deployment
+To deploy our application, we setup the bash script as shown below with comments for each command:  
+
```
Comment (Ericgacoki) [inline content/articles/spring-cloud-k8/index.md:180]:
```suggestion
To deploy our application, we setup a `bash script` as explained below:  

```

File: content/articles/spring-cloud-k8/index.md [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+# run checks on pods to check if they are indeed running
+kubectl get pods
+
+```
+
+### Conclusion
+In this tutorial, we've seen how we can set up a Spring Cloud Kubernetes project.
+We've seen how we can use both tools to set up a microservice platform application with each providing its best `power`.
+
+The full source code is available [here](https://replit.com/@odiwuoramos/spring-cloud#kubernetes-guide/client-service/client-config.yaml)
```
Comment (Ericgacoki) [inline content/articles/spring-cloud-k8/index.md:223]:
```suggestion
In this tutorial, we've seen how we can set up a Spring Cloud project and use Spring Boot and Kubernetes to integrate a micro-service platform application with each providing great performance.

The full source code can be found [here](https://replit.com/@odiwuoramos/spring-cloud#kubernetes-guide/client-service/client-config.yaml).
```

**Review by Ericgacoki — CHANGES_REQUESTED** [review]
Hey @odiwuoramos, I've completed `preliminary review`🚀. This is absolutely an enlightening tutorial😎. However, I've made a few suggestions to improve the flow.
All you need to do is to `commit the suggestions` from this panel. At the moment, Please `DO NOT` manually edit the file as doing so will mess up the suggestions by resolving them incorrectly.
Ping me when you're done so we may proceed to `tech-review`.

**Comment by odiwuoramos** [comment]
@Ericgacoki , all changes done

**Comment by Ericgacoki** [comment]
> @Ericgacoki , all changes done

Not yet! We still have some unresolved... Expand the `hidden` section to see them.

**Comment by odiwuoramos** [comment]
> > @Ericgacoki , all changes done
> 
> Not yet! We still have some unresolved... Expand the `hidden` section to see them.

updated! thanks for heads up

**Review by Ericgacoki — APPROVED** [review]
Good work @odiwuoramos, now heading to `tech review`. 
@ahmadmardeni1, kindly update the labels. Thanks.

**Comment by ahmadmardeni1** [comment]
Please only commit when you are sure that you are ready - one commit instead of multiple commits. @odiwuoramos

**Comment by Ericgacoki** [comment]
> Please only commit when you are sure that you are ready - one commit instead of multiple commits. @odiwuoramos

@ahmadmardeni1,
The commits were from the suggestions I made. I still don't have the permission to edit files directly. Meanwhile, this is the only way to apply my changes.

**Comment by ahmadmardeni1** [comment]
I know - It is alright. But I mean he is making a lot of commits instead of one (take a look at the screenshot).

Capture

**Comment by Ericgacoki** [comment]
@odiwuoramos,
[This GitHub doc](https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request) can help you learn how to quickly incorporate more than one suggested change into a **single commit**.

regards @ahmadmardeni1.

**Review by Ericgacoki — CHANGES_REQUESTED** [review]
Greetings @odiwuoramos, I'm done with the `tech review`.
|Formating| Code Testing | Tutorial Flow|
|--|--|--|
| ✅ Good | ✅ Passed| ✅ Good |

 Please check out the following minor suggestions that I've made.

File: content/articles/spring-cloud-k8/index.md [hunk 7]
```diff
@@ -0,0 +1,219 @@
+### Tutorial overview
+In this tutorial, we will learn how to use Kubernetes and Spring Cloud to build a micro-service application and integrate it with Spring Boot.
```
Comment (Ericgacoki) [inline content/articles/spring-cloud-k8/index.md:2]:
```suggestion
---
layout: engineering-education
status: publish
published: true
url: /spring-cloud-k8/
title: Guide to Spring Cloud Kubernetes (K8s)
description: This tutorial will walk you through the process of creating a Spring Cloud Kubernetes application.  Spring Cloud is a Spring module that offers RAD (Rapid Application Development) functionality to the Spring framework.
author: odiwuor-amos
date: 2021-07-23T00:00:00-10:00
topics: []
excerpt_separator: <!--more-->
images:

  - url: /engineering-education/spring-cloud-k8/hero.png
    alt: Spring Cloud K8 image
---

In this tutorial, we will learn how to use Kubernetes and Spring Cloud to build a micro-service application and integrate it with Spring Boot. Spring Cloud is a Spring module that offers RAD (Rapid Application Development) functionality to the Spring framework.
<!--more-->
With the assistance of the Spring Cloud Framework, we can easily create cloud-based allocation.
```
Added Front matter.

File: content/articles/spring-cloud-k8/index.md [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+# run checks on pods to check if they are indeed running
+kubectl get pods
+
+```
+
+### Conclusion
+In this tutorial, we've seen how we can set up a Spring Cloud project and use Spring Boot and Kubernetes to integrate a micro-service platform application with each providing great performance.
+
+The full source code can be found [here](https://replit.com/@odiwuoramos/spring-cloud#kubernetes-guide/client-service/client-config.yaml).
```
Comment (Ericgacoki) [inline content/articles/spring-cloud-k8/index.md:219]:
```suggestion
The full source code can be found [here](https://replit.com/@odiwuoramos/spring-cloud#kubernetes-guide/client-service/client-config.yaml).

Happy coding!

---
Peer Review Contributions by: [Eric gacoki](/engineering-education/authors/eric-gacoki/)
```
Added peer reviewer line.

**Review by Ericgacoki — CHANGES_REQUESTED** [review]
I've added:
- [x] Front matter and 
- [x] Peer reviewer line.

**Comment by odiwuoramos** [comment]
Hi, @Ericgacoki I've made all the requested changes, I've learnt a lot from you!

**Comment by Ericgacoki** [comment]
Great work @odiwuoramos, this PR will be `ready for final review` after passing the `plagiarism check`. Thank you for your cooperation 🚀.

cc: @ninjaginja, @ahmadmardeni1.

**Review by Ericgacoki — APPROVED** [review]

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by hectorkambow** [comment]
Great work gents @odiwuoramos 

cc @Ericgacoki 

very minor edits for flow/format - please review my changes to improve future articles.

**Pull request merged**